### PR TITLE
[docs] Improve discoverability of Filtering section of docs

### PR DESCRIPTION
**Pull request by MBilalShafi**
Related to #9034.

Divide into sections which can be extended further by adding sections like Examples/recipes.

image

https://deploy-preview-9074--material-ui-x.netlify.app/x/react-data-grid/filtering/server-side/

### Pros:

1. Less clutter
2. Easy discoverability of the section the user is searching (let's say I'm looking for Quick filter only, why skim through the whole document)
3. Opportunity to differentiate commercial features with MIT.
4. Opportunity to elaborate more on some sections or add new sections (e.g adding another demo to server-side filtering or adding a new `Examples` section for recipes)
5. Opportunity to showcase new features (like Header filters) better

### Cons:

1. Not sure if two-level nesting is a problem (haven't seen it for any other section)

Any suggestions are welcome

### CODE REVIEW

**Review by romgrk — APPROVED** [review]
LGTM, love the change, some of our docs pages are huge.

File: docs/data/pages.ts [hunk 1]
```diff
@@ -64,7 +64,22 @@ const pages: MuiPage[] = [
       },
       { pathname: '/x/react-data-grid/editing' },
       { pathname: '/x/react-data-grid/sorting' },
-      { pathname: '/x/react-data-grid/filtering' },
+      {
+        pathname: '/x/react-data-grid/filtering',
+        title: 'Filtering',
+        children: [
+          { pathname: '/x/react-data-grid/filtering/overview' },
+          { pathname: '/x/react-data-grid/filtering/customization' },
```
Comment (oliviertassinari) [inline docs/data/pages.ts:72]: Would it be clearer?

```suggestion
          { pathname: '/x/react-data-grid/filtering/customization', title: 'Filter customization' },
```

**Review by oliviertassinari — COMMENTED** [review]
This change makes sense 👍

File: docs/data/pages.ts (hunk 1 above)
Comment (MBilalShafi) [reply docs/data/pages.ts:75]: IMO it'd be obvious as it's under `Filtering` section anyways.

File: docs/data/data-grid/filtering/quick-filter.md [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    searchInput.split(',').map((value) => value.trim())
+  }
+  quickFilterFormatter={(quickFilterValues) => quickFilterValues.join(', ')}
+  debounceMs={200} // time before applying the new quick filter value
+/>
+```
+
+In the following demo, the quick filter value `"Saint Martin, Saint Lucia"` will return rows with country is Saint Martin or Saint Lucia.
+
+{{"demo": "QuickFilteringCustomizedGrid.js", "bg": "inline", "defaultCodeOpen": false}}
```
Comment (cherniavskii) [inline docs/data/data-grid/filtering/quick-filter.md:100]:
```suggestion
{{"demo": "QuickFilteringCustomizedGrid.js", "bg": "inline", "defaultCodeOpen": false}}

## API

- [DataGrid](/x/api/data-grid/data-grid/)
- [DataGridPro](/x/api/data-grid/data-grid-pro/)
- [DataGridPremium](/x/api/data-grid/data-grid-premium/)
```

We usually link at least these API pages at the bottom of the page.

Side note: For this page specifically, we might want to generate a doc page for the `GridToolbarQuickFilter` component (similarly to https://mui.com/x/api/data-grid/grid-filter-panel/).

File: docs/data/data-grid/filtering/server-side.md [hunk 3]
```diff
@@ -0,0 +1,9 @@
+# Data Grid - Server-side filter
+
+<p class="description">Filter directly from the server.</p>
+
+Filtering can be run server-side by setting the `filterMode` prop to `server`, and implementing the `onFilterModelChange` handler.
+
+The example below demonstrates how to achieve server-side filtering.
+
+{{"demo": "ServerFilterGrid.js", "bg": "inline"}}
```
Comment (cherniavskii) [inline docs/data/data-grid/filtering/server-side.md:9]:
```suggestion
{{"demo": "ServerFilterGrid.js", "bg": "inline"}}

## API

- [DataGrid](/x/api/data-grid/data-grid/)
- [DataGridPro](/x/api/data-grid/data-grid-pro/)
- [DataGridPremium](/x/api/data-grid/data-grid-premium/)
```

**Review by cherniavskii — COMMENTED** [review]
Awesome, looks much better now! 🎉

File: docs/data/data-grid/filtering/customization.md [hunk 4]
```diff
@@ -1,119 +1,6 @@
-# Data Grid - Filtering
```
Comment (cherniavskii) [inline docs/data/data-grid/filtering/customization.md:1]: > Cons:
> 1. Not sure if two-level nesting is a problem (haven't seen it for any other section)

Do you mean `Data Grid > Filtering > ...`? We already do this for Columns, Rows, Selection pages

File: docs/data/data-grid/filtering/index.md [hunk 5]
```diff
@@ -0,0 +1,50 @@
+# Data Grid - Filtering
```
[the remark this answers is not in the document]
Comment (MBilalShafi) [reply docs/data/data-grid/filtering/index.md:1]: That's a nice idea 👌

File: docs/data/data-grid/filtering/multi-filters.md [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+This demo implements a basic use case to prevent showing multiple filters for one column.
+
+{{"demo": "DisableMultiFiltersDataGridPro.js", "bg": "inline", "defaultCodeOpen": false}}
+
+## Disable action buttons [<span class="plan-pro"></span>](/x/introduction/licensing/#pro-plan)
+
+To disable `Add filter` or `Remove all` buttons, pass `disableAddFilterButton` or `disableRemoveAllButton` to `componentsProps.filterPanel`.
+
+{{"demo": "DisableActionButtonsDataGridPro.js", "bg": "inline", "defaultCodeOpen": false}}
```
Comment (cherniavskii) [inline docs/data/data-grid/filtering/multi-filters.md:27]:
```suggestion
{{"demo": "DisableActionButtonsDataGridPro.js", "bg": "inline", "defaultCodeOpen": false}}

## API

- [GridFilterPanel](/x/api/data-grid/grid-filter-panel/)
- [DataGrid](/x/api/data-grid/data-grid/)
- [DataGridPro](/x/api/data-grid/data-grid-pro/)
- [DataGridPremium](/x/api/data-grid/data-grid-premium/)
```

File: docs/data/data-grid/filtering/customization.md (hunk 4 above)
Comment (MBilalShafi) [reply docs/data/data-grid/filtering/customization.md:1]: I mean the path set in [pages.ts](https://github.com/mui/mui-x/blob/5172a58a68e794dc9da87e15d292cf434567fc83/docs/data/pages.ts) file.

It's always something like `/x/package-name/feature-name` but in this PR, I am using something like `/x/package-name/feature-name/sub-feature-name`

Comment (cherniavskii) [reply docs/data/data-grid/filtering/customization.md]: I don't foresee any issues with the `/feature-name/sub-feature-name` URL, and it looks nicer to me 👍

When I was splitting the Editing page in https://github.com/mui/mui-x/pull/7772, I didn't think much about URLs.
My main motivation was to keep the demos of each subpage in a separate folder (`docs/data/data-grid/*`).
Currently, all filtering-related demos are in the same folder, which makes it a bit harder to navigate (but easier to review this PR 😅).
It's not a big deal though.

File: docs/data/data-grid/filtering/multi-filters.md (hunk 6 above)
Comment (MBilalShafi) [reply docs/data/data-grid/filtering/multi-filters.md:27]: Done ✅

**Review by cherniavskii — APPROVED** [review]

**Comment by MBilalShafi** [comment]
Need to merge https://github.com/mui/material-ui/pull/37455 to fix the build

File: docs/data/data-grid/state/state.md [hunk 7]
```diff
@@ -11,7 +11,7 @@ This prop has the same format as the returned value of `apiRef.current.exportSta
 The `initialState` can only be used to set the initial value of the state.
 The Data Grid will not react if you change the `initialState` value later on.
 
-If you need to fully control specific models, use the control props instead (e.g. [`prop.filterModel`](/x/react-data-grid/filtering/#controlled-filters) or [`prop.sortModel`](https://mui.com/x/react-data-grid/sorting/#controlled-sort-model)).
+If you need to fully control specific models, use the control props instead (e.g. [`prop.filterModel`](/x/react-data-grid/filtering/customization/#controlled-filters) or [`prop.sortModel`](https://mui.com/x/react-data-grid/sorting/#controlled-sort-model)).
```
Comment (cherniavskii) [inline docs/data/data-grid/state/state.md:14]:
```suggestion
If you need to fully control specific models, use the control props instead (e.g. [`prop.filterModel`](/x/react-data-grid/filtering/#controlled-filters) or [`prop.sortModel`](https://mui.com/x/react-data-grid/sorting/#controlled-sort-model)).
```

File: docs/data/data-grid/column-spanning/column-spanning.md [hunk 8]
```diff
@@ -25,7 +25,7 @@ When using `colSpan`, some other features may be pointless or may not work as ex
 To avoid confusing grid layout, consider disabling the following features for the column(s) that are affected by `colSpan`:
 
 - [sorting](/x/react-data-grid/sorting/#disable-the-sorting)
-- [filtering](/x/react-data-grid/filtering/#disable-the-filters)
+- [filtering](/x/react-data-grid/filtering/customization/#disable-the-filters)
```
Comment (cherniavskii) [inline docs/data/data-grid/column-spanning/column-spanning.md:28]:
```suggestion
- [filtering](/x/react-data-grid/filtering/#disable-the-filters)
```

File: docs/pages/x/api/data-grid/grid-filter-model.md [hunk 9]
```diff
@@ -7,7 +7,7 @@
 :::info
 For examples and details on the usage, check the following pages:
 
-- [Pass filters to the grid](/x/react-data-grid/filtering/#pass-filters-to-the-data-grid)
+- [Pass filters to the grid](/x/react-data-grid/filtering/customization/#pass-filters-to-the-data-grid)
```
Comment (cherniavskii) [inline docs/pages/x/api/data-grid/grid-filter-model.md:10]:
```suggestion
- [Pass filters to the grid](/x/react-data-grid/filtering/#pass-filters-to-the-data-grid)
```

File: packages/grid/x-data-grid/src/models/gridFilterModel.ts [hunk 10]
```diff
@@ -3,7 +3,7 @@ import { GridFilterItem, GridLogicOperator } from './gridFilterItem';
 /**
  * Model describing the filters to apply to the grid.
  * @demos
- *   - [Pass filters to the grid](/x/react-data-grid/filtering/#pass-filters-to-the-data-grid)
+ *   - [Pass filters to the grid](/x/react-data-grid/filtering/customization/#pass-filters-to-the-data-grid)
```
Comment (cherniavskii) [inline packages/grid/x-data-grid/src/models/gridFilterModel.ts:6]:
```suggestion
 *   - [Pass filters to the grid](/x/react-data-grid/filtering/#pass-filters-to-the-data-grid)
```

**Review by cherniavskii — COMMENTED** [review]
Some of the errors reported by link-check are valid.
For the one that is being used in monorepo, you can run `yarn docs:link-check` locally and commit the produced changes.

**Comment by MBilalShafi** [comment]
> you can run yarn docs:link-check locally and commit the produced changes.

Thanks @cherniavskii for the tip-off, I didn't know this fixes the build 🙏

File: docs/data/pages.ts (hunk 1 above)
Comment (oliviertassinari) [reply docs/data/pages.ts:75]: Yeah, OK, it makes sense

**Pull request merged**